File: web/src/data/LastSyncStorage.ts
import { Timestamp } from 'firebase/firestore'
import type { Channel } from './UserDocSignal'

/**
 * Per-device, per-user, per-channel watermark for the delta-pull mechanism.
 * Stored as ms-since-epoch in localStorage. See `docs/live-cross-platform-sync.md`.
 * Missing value → returns Timestamp(0, 0) → next pull is a full pull.
 */

function key(channel: Channel, userId: string): string {
  return `lastSync_${channel}_${userId}`
}

function safeLocalStorage(): Storage | null {
  if (typeof localStorage === 'undefined') return null
  return localStorage
}

export const LastSyncStorage = {
  /** Read the watermark, or `Timestamp(0, 0)` if unset. */
  read(channel: Channel, userId: string): Timestamp {
    const ls = safeLocalStorage()
    if (!ls) return new Timestamp(0, 0)
    const raw = ls.getItem(key(channel, userId))
    if (raw == null) return new Timestamp(0, 0)
    const ms = Number(raw)
    if (!Number.isFinite(ms) || ms <= 0) return new Timestamp(0, 0)
    const seconds = Math.floor(ms / 1000)
    const nanos = (ms % 1000) * 1_000_000
    return new Timestamp(seconds, nanos)
  },

  /** Caller must invoke only after a successful apply, so a transient
   *  failure can't advance past the real `max(updatedAt)` the device has
   *  actually persisted. */
  write(channel: Channel, userId: string, timestamp: Timestamp): void {
    const ls = safeLocalStorage()
    if (!ls) return
    const ms = timestamp.seconds * 1000 + Math.floor(timestamp.nanoseconds / 1_000_000)
    ls.setItem(key(channel, userId), String(ms))
  },

  /** Used by the full-repair pull (count() detection) to start over. */
  clear(channel: Channel, userId: string): void {
    const ls = safeLocalStorage()
    if (!ls) return
    ls.removeItem(key(channel, userId))
  },
}

File: app/src/main/java/org/alkaline/taskbrain/data/LastSyncStorage.kt
package org.alkaline.taskbrain.data

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.google.firebase.Timestamp

/**
 * Per-device, per-user watermark for the delta-pull mechanism. Stored as
 * ms-since-epoch in SharedPreferences. See `docs/live-cross-platform-sync.md`.
 * Missing value → returns Timestamp(0, 0) → next pull is a full pull.
 */
object LastSyncStorage {
    private const val TAG = "LastSyncStorage"
    private const val PREFS_NAME = "taskbrain_prefs"

    private var prefs: SharedPreferences? = null

    /** Hook up persistence. Idempotent. Call once from Application.onCreate. */
    fun attach(context: Context) {
        if (prefs != null) return
        prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    private fun key(channel: UserDocSignal.Channel, userId: String): String =
        "lastSync_${channel.name}_$userId"

    /** Read the watermark for ([channel], [userId]), or `Timestamp(0, 0)` if unset. */
    fun read(channel: UserDocSignal.Channel, userId: String): Timestamp {
        val p = prefs
        if (p == null) {
            Log.w(TAG, "read($channel,$userId) called before attach(); returning 0")
            return Timestamp(0, 0)
        }
        val ms = p.getLong(key(channel, userId), 0L)
        if (ms <= 0L) return Timestamp(0, 0)
        val seconds = ms / 1000
        val nanos = ((ms % 1000) * 1_000_000).toInt()
        return Timestamp(seconds, nanos)
    }

    /** Caller must invoke only after a successful apply, so a transient
     *  failure can't advance past the real `max(updatedAt)` the device has
     *  actually persisted. */
    fun write(channel: UserDocSignal.Channel, userId: String, timestamp: Timestamp) {
        val p = prefs
        if (p == null) {
            Log.w(TAG, "write($channel,$userId) called before attach(); dropping")
            return
        }
        val ms = timestamp.seconds * 1000L + timestamp.nanoseconds / 1_000_000L
        p.edit().putLong(key(channel, userId), ms).apply()
    }

    /** Used by the full-repair pull (count() detection) to start over. */
    fun clear(channel: UserDocSignal.Channel, userId: String) {
        prefs?.edit()?.remove(key(channel, userId))?.apply()
    }
}
